src/algebra/group/type_tags.lean

```diff
@@ -27,14 +27,54 @@ def additive (α : Type*) := α
 multiplicative structure. -/
 def multiplicative (α : Type*) := α
 
-instance [inhabited α] : inhabited (additive α) := ⟨(default _ : α)⟩
-instance [inhabited α] : inhabited (multiplicative α) := ⟨(default _ : α)⟩
+/-- Reinterpret `x : α` as an element of `additive α`. -/
+def additive.of_mul (x : α) : additive α := x
+
+/-- Reinterpret `x : additive α` as an element of `α`. -/
+def additive.to_mul (x : additive α) : α := x
+
+lemma of_mul_inj : function.injective (@additive.of_mul α) := λ _ _, id
+lemma to_mul_inj : function.injective (@additive.to_mul α) := λ _ _, id
+
+/-- Reinterpret `x : α` as an element of `multiplicative α`. -/
+def multiplicative.of_add (x : α) : multiplicative α := x
+
+/-- Reinterpret `x : multiplicative α` as an element of `α`. -/
+def multiplicative.to_add (x : multiplicative α) : α := x
+
+lemma of_add_inj : function.injective (@multiplicative.of_add α) := λ _ _, id
+lemma to_add_inj : function.injective (@multiplicative.to_add α) := λ _ _, id
+
+@[simp] lemma to_add_of_add (x : α) : (multiplicative.of_add x).to_add = x := rfl
+@[simp] lemma of_add_to_add (x : multiplicative α) : multiplicative.of_add x.to_add = x := rfl
+
+@[simp] lemma to_mul_of_mul (x : α) : (additive.of_mul x).to_mul = x := rfl
+@[simp] lemma of_mul_to_mul (x : additive α) : additive.of_mul x.to_mul = x := rfl
+
+instance [inhabited α] : inhabited (additive α) := ⟨additive.of_mul (default α)⟩
+instance [inhabited α] : inhabited (multiplicative α) := ⟨multiplicative.of_add (default α)⟩
 
 instance additive.has_add [has_mul α] : has_add (additive α) :=
-{ add := ((*) : α → α → α) }
+{ add := λ x y, additive.of_mul (x.to_mul * y.to_mul) }
 
 instance [has_add α] : has_mul (multiplicative α) :=
-{ mul := ((+) : α → α → α) }
+{ mul := λ x y, multiplicative.of_add (x.to_add + y.to_add) }
+
+@[simp] lemma of_add_add [has_add α] (x y : α) :
+  multiplicative.of_add (x + y) = multiplicative.of_add x * multiplicative.of_add y :=
+rfl
+
+@[simp] lemma to_add_mul [has_add α] (x y : multiplicative α) :
+  (x * y).to_add = x.to_add + y.to_add :=
```

Comment:
@robertylewis Can we hook this up to `norm_cast` even though it's not (and should not be) `coe`?

Reply:
Not really. It might work with the current `norm_cast`, but in the open PR (that will be finished eventually), we classify lemmas according to the position of `coe`. It's not really designed to work with arbitrary coercion operators.